File: iosApp/Services/EditHistory.swift
import Foundation

class EditHistory {

    private var undoStack: [EditParameters] = []
    private var redoStack: [EditParameters] = []

    var canUndo: Bool { undoStack.count > 1 }
    var canRedo: Bool { !redoStack.isEmpty }
    var undoCount: Int { undoStack.count }
    var redoCount: Int { redoStack.count }

    func push(_ parameters: EditParameters) {
        undoStack.append(parameters)
        redoStack.removeAll()
    }

    @discardableResult
    func undo() -> EditParameters? {
        guard undoStack.count > 1, let last = undoStack.popLast() else { return nil }
        redoStack.append(last)
        return undoStack.last
    }

    @discardableResult
    func redo() -> EditParameters? {
        guard let last = redoStack.popLast() else { return nil }
        undoStack.append(last)
        return last
    }
}

File: shared/src/commonMain/kotlin/com/example/photoeditor/EditHistory.kt
package com.example.photoeditor

class EditHistory {
    private val undoStack = ArrayDeque<EditParameters>()
    private val redoStack = ArrayDeque<EditParameters>()

    val canUndo: Boolean get() = undoStack.isNotEmpty()
    val canRedo: Boolean get() = redoStack.isNotEmpty()

    fun push(parameters: EditParameters) {
        undoStack.addLast(parameters)
        redoStack.clear()
    }

    fun undo(): EditParameters? {
        if (undoStack.isEmpty()) return null
        val current = undoStack.removeLast()
        redoStack.addLast(current)
        return undoStack.lastOrNull()
    }

    fun redo(): EditParameters? {
        if (redoStack.isEmpty()) return null
        val next = redoStack.removeLast()
        undoStack.addLast(next)
        return next
    }

    fun clear() {
        undoStack.clear()
        redoStack.clear()
    }
}
